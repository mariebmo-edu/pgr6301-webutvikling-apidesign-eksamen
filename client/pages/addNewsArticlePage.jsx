import { useContext, useState } from "react";
import { NewsApiContext } from "../context/newsApiContext";
import { useNavigate } from "react-router-dom";
import { FormInput } from "../components/modules/formInput";
import { ErrorState } from "../components/states/errorState";
import { useLoading } from "../utils/loadingUtils";

export function AddNewsArticlePage({ userName }) {
  const { createNewsArticle, getNewsArticle } = useContext(NewsApiContext);

  const [title, setTitle] = useState("");
  const [authors, setAuthors] = useState([userName]);
  const [image, setImage] = useState("");
  const [categories, setCategories] = useState([]);
  const [content, setContent] = useState("");
  const [errorMsg, setErrorMsg] = useState("");

  const navigate = useNavigate();
  const { loading, error, data } = useLoading(
    async () => getNewsArticle({ paramTitle: title }),
    [title]
  );

  async function handleSubmit(event) {
    event.preventDefault();

    if (title !== "" && authors !== [] && categories !== [] && content !== "") {
      if (data && data.title === title) {
        setErrorMsg("Title already exists");
        setTitle("");
      } else if (!data || data.title !== title) {
        await createNewsArticle({ title, authors, image, categories, content });
        navigate("/");
      }
    } else {
      setErrorMsg("You need input in all fields.");
    }
  }

  return (
    <div>
      {errorMsg && <ErrorState error={errorMsg} />}
      <form onSubmit={handleSubmit}>
        <FormInput
          inputId={"title-input"}
          type={"text"}
          label={"Title:"}
          value={title}
          onChangeValue={setTitle}
        />
        <FormInput
          inputId={"image-input"}
          type={"text"}
          label={"Image Url:"}
          value={image}
          onChangeValue={setImage}
        />
        <FormInput
          inputId={"category-input"}
          type={"category-checkbox"}
          label={"Category"}
          value={categories}
          onChangeValue={setCategories}
        />
        <FormInput
          inputId={"content-input"}
          type={"text"}
          label={"Content:"}
          value={content}
          onChangeValue={setContent}
        />
        <button>Submit</button>
      </form>
    </div>
  );
}
