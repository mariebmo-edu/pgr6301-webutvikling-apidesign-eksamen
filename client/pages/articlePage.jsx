import { useContext } from "react";
import { NewsApiContext } from "../context/newsApiContext";
import { useLoading } from "../utils/loadingUtils";
import { LoadingState } from "../components/states/loadingState";
import { ErrorState } from "../components/states/errorState";

export function ArticlePage() {
  const params = new Proxy(new URLSearchParams(window.location.search), {
    get: (searchParams, prop) => searchParams.get(prop),
  });

  const paramTitle = params.title;

  const { getNewsArticle } = useContext(NewsApiContext);
  const { loading, error, data } = useLoading(async () =>
    getNewsArticle({ paramTitle })
  );

  if (loading) {
    return <LoadingState />;
  }

  if (error) {
    return (
      <>
        <ErrorState error={error} />
      </>
    );
  }

  return (
    <div className={"window-content-container"}>
      <img src={data.image} alt={"header image of article"} />
      <h1>{data.title}</h1>
      <p>{data.content}</p>
    </div>
  );
}
