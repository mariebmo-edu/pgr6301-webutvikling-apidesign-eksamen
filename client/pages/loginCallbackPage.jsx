import { useContext, useEffect, useState } from "react";
import { WaitingState } from "../components/states/waitingState";
import { useNavigate, useParams } from "react-router-dom";
import { ErrorState } from "../components/states/errorState";
import { LoginApiContext } from "../context/loginApiContext";

export function LoginCallbackPage({ reload, config }) {
  const { provider } = useParams();
  const [error, setError] = useState();
  const navigate = useNavigate();
  const { registerLogin } = useContext(LoginApiContext);

  //Login is based on my mock-exam, which is heavily influenced by the lecture 12 reference
  // I've added comments to show I understand how the code works.

  //useEffect is called once - making sure it doesn't run for ever.
  useEffect(async () => {
    //gets the different tokens/errors/states/code from the search parameters. Substrind to remove the first ?
    const { access_token, error, error_description, state, code } =
      Object.fromEntries(
        new URLSearchParams(window.location.hash.substring(1))
      );

    //the expected_state is set when clicking the login-button, being a random string of 50 characters. It is made to
    //make sure the request is sent from inside the application - and not elsewhere.
    const expected_state = window.sessionStorage.getItem("expected_state");
    if (!state || expected_state !== state) {
      setError("Unexpected State");
      return;
    }

    //if there is an error from the params, we're setting them here and exiting.
    if (error || error_description) {
      setError(`Error: ${error} ${error_description}`);
      return;
    }

    //if you got a code - we'll try to make it into a token!
    // We're getting the config from the login-page, and selecting the endpoints and client_id for that provider.
    if (code) {
      const { client_id, token_endpoint } = config[provider];

      //we're getting the code_verifier from the local session storage, again making sure it is the same person the whole time.
      const code_verifier = window.sessionStorage.getItem("code_verifier");

      //getting the payload ready to send to the token_endpoint to retrieve the final token
      const payload = {
        grant_type: "authorization_code",
        code,
        client_id,
        code_verifier,
        redirect_uri: window.location.origin + "/login/microsoft/callback",
      };

      //sending a post-request to the endpoint with the payload above - waiting to get a token back
      const res = await fetch(token_endpoint, {
        method: "POST",
        body: new URLSearchParams(payload),
      });

      //if the response is not ok, we're setting an error and exiting.
      if (!res.ok) {
        setError(`Failed to fetch token, ${res.status} ${res.statusText}`);
        return;
      }

      //if everything is STILL going well, we're getting the access-token from the response. We're then trying to register the
      //login at our side, sending the provider and token to the API.
      const { access_token } = await res.json();
      await registerLogin(provider, { access_token });

      //WOO! Everything went as planned - we're going back to the main page, and reloading. Job done.
      navigate("/");
      reload();
      return;
    }

    //There's no access token. ERROR! exit.
    if (!access_token) {
      setError("Missing access_token");
      return;
    }

    //you got an access-token? We'll give you a shot! Let's register you!
    await registerLogin(provider, { access_token });
    reload();
    navigate("/");
  }, []);

  //Oopsie, error. We'll return an error-state
  if (error) {
    return (
      <>
        <ErrorState error={error} />
      </>
    );
  }

  //You'll either get a redirect or an error in a second, but just in case, here's some waiting music.
  return <WaitingState />;
}
