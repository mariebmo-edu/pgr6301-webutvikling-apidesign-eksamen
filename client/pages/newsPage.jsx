import { useContext, useEffect, useState } from "react";
import { NewsApiContext } from "../context/newsApiContext";
import { useLoading } from "../utils/loadingUtils";
import { CompactNewsCard } from "../components/modules/compactNewsCard";
import { LoadingState } from "../components/states/loadingState";
import { ErrorState } from "../components/states/errorState";
import { FormInput } from "../components/modules/formInput";
import { Link } from "react-router-dom";

export function NewsPage() {
  const { listNews } = useContext(NewsApiContext);

  const [category, setCategory] = useState("");
  const [categoryQuery, setCategoryQuery] = useState("");
  const [reloadNews, setReloadNews] = useState(false);

  const { loading, error, data } = useLoading(
    async () => await listNews({ category }),
    [category, reloadNews]
  );

  const [webSocket, setWebSocket] = useState();

  useEffect(() => {
    const webSocket = new WebSocket(window.location.origin.replace(/^http/, "ws"));

    webSocket.onopen = (event) => {
      console.log("The socket is open");
    };

    webSocket.onmessage = (event) => {
      console.log("message", event.data);
      setReloadNews(!reloadNews);
    };
    setWebSocket(webSocket);
  }, []);

  function handleSubmitQuery(event) {
    event.preventDefault();
    setCategory(categoryQuery);
  }

  return (
    <>
      <div className={"side-bar-container secondary-background"}>
        <form onSubmit={handleSubmitQuery}>
          <FormInput
            id="news-category-dropdown"
            inputId={"categories"}
            value={category}
            label={"Categories:"}
            type={"category-dropdown"}
            onChangeValue={setCategoryQuery}
          />
          <button>Filter</button>
        </form>
      </div>

      <div className={"window-content-container"}>
        {loading && <LoadingState key={"loading-component"} />}
        {error && <ErrorState error={error} key={"error-component"} />}
        {data && (
          <div className={"flex-container"} key={"data-component"}>
            {data.map((news) => (
              <Link to={`article?title=${news.title}`} key={news.title}>
                <CompactNewsCard news={news} />
              </Link>
            ))}
          </div>
        )}
      </div>
    </>
  );
}
