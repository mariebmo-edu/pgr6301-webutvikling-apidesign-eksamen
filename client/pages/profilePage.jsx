import { ErrorState } from "../components/states/errorState";

export function ProfilePage({ user }) {
  if (Object.keys(user)[0] === "microsoft") {
    user = user["microsoft"];
    user.picture =
      "https://www.itdp.org/wp-content/uploads/2021/06/avatar-man-icon-profile-placeholder-260nw-1229859850-e1623694994111.jpg";
    user.authority = "Admin";
  } else if (Object.keys(user)[0] === "google") {
    user = user["google"];
    user.authority = "User";
  } else {
    return (
      <ErrorState
        error={"Your profile is not recognized. Have you logged in?"}
      />
    );
  }

  return (
    <>
      <h1>{user.name}</h1>
      <h2>{user.authority}</h2>
      <img src={user.picture} alt={"profile picture of " + user.name} />
      <p>Email: {user.email}</p>
    </>
  );
}
