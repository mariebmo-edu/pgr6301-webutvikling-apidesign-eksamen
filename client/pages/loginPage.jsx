import { useContext, useEffect } from "react";
import { Route, Routes, useNavigate } from "react-router-dom";
import { LoginCallbackPage } from "./loginCallbackPage";
import { LoginApiContext } from "../context/loginApiContext";
import { WaitingState } from "../components/states/waitingState";
import { LoginButton } from "../components/modules/loginButton";

//Login is based on my mock-exam, which is heavily influenced by the lecture 12 reference
// I've added comments to show I understand how the code works.

export function EndSession({ reload }) {
  //removes the cookies (endSession) and redirects user to root.
  const navigate = useNavigate();
  const { endSession } = useContext(LoginApiContext);
  useEffect(async () => {
    await endSession();
    reload();
    navigate("/");
  });

  return <WaitingState />;
}

export function StartLogin({ config }) {
  return (
    <div className={"center-buttons"}>
      <LoginButton
        image={
          "https://www.freepnglogos.com/uploads/google-logo-png/google-logo-png-webinar-optimizing-for-success-google-business-webinar-13.png"
        }
        config={config}
        provider={"google"}
      />
      <br />
      <LoginButton
        image={
          "https://img.cppng.com/download/2020-06/28348-7-microsoft-logo-hd.png"
        }
        config={config}
        provider={"microsoft"}
      />
    </div>
  );
}

export function LoginPage({ config, reload }) {
  //returns routes connected to the different components inside LoginPage.
  //Start Login -> Buttons to take you to the corresponding OpenId provider
  //provider/callback -> The landing page after logging in, handles codes -> token and token -> cookie-exchange
  //endsession -> logs the user out
  //Start Login (again) - If you access any other link after /login, you are redirected to StartLogin.

  return (
    <div className={"window-content-container"}>
      <Routes>
        <Route path={"/"} element={<StartLogin config={config} />} />
        <Route
          path={"/:provider/callback"}
          element={<LoginCallbackPage config={config} reload={reload} />}
        />
        <Route path={"/endsession"} element={<EndSession reload={reload} />} />
        <Route path={"*"} element={<StartLogin config={config} />} />
      </Routes>
    </div>
  );
}
