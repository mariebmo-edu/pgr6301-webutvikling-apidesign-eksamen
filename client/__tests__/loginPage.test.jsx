import ReactDOM from "react-dom";
import { BrowserRouter as Router } from "react-router-dom";
import { LoginPage } from "../pages/loginPage";

describe("Log in component", () => {
  it("shows login options", () => {
    const domElement = document.createElement("div");
    const config = {};
    const reload = {};
    ReactDOM.render(
      <Router>
        <LoginPage config={config} reload={reload} />
      </Router>,
      domElement
    );

    expect(domElement.innerHTML).toMatchSnapshot();
    expect(
      Array.from(domElement.querySelectorAll(".login-button-text")).map(
        (element) => element.innerHTML
      )
    ).toEqual(["Log in with google", "Log in with microsoft"]);
  });
});
