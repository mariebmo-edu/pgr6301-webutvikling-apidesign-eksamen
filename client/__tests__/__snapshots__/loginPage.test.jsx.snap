// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Log in component shows login options 1`] = `"<div class=\\"window-content-container\\"><div class=\\"center-buttons\\"><button class=\\"login-button\\"><img class=\\"button-image\\" src=\\"https://www.freepnglogos.com/uploads/google-logo-png/google-logo-png-webinar-optimizing-for-success-google-business-webinar-13.png\\" alt=\\"login with google\\"><span class=\\"login-button-text\\">Log in with google</span></button><br><button class=\\"login-button\\"><img class=\\"button-image\\" src=\\"https://img.cppng.com/download/2020-06/28348-7-microsoft-logo-hd.png\\" alt=\\"login with microsoft\\"><span class=\\"login-button-text\\">Log in with microsoft</span></button></div></div>"`;
